# ■ 자연어 처리 (NLP)
Natural Language Processing 이란

**자연어 처리(NLP, Natural Language Processing)** 는 컴퓨터가 인간의 언어를 이해하고, 해석하고, 생성하는 기술을 말합니다. NLP는 언어학과 컴퓨터 과학의 융합 분야로, 언어를 처리하고 분석하는 다양한 작업을 포함합니다. 그 목적은 사람과 컴퓨터 간의 상호작용을 원활하게 만들어 주는 것입니다.

자연어 처리의 주요 작업은 다음과 같습니다:

1. **형태소 분석 (Morphological Analysis)**: 문장에서 단어를 나누고, 각 단어의 형태를 분석합니다. 예를 들어, 한국어에서 "사랑한다"라는 단어는 "사랑"과 "한다"로 나누어집니다.

1. **구문 분석 (Syntactic Parsing)**: 문장의 구조를 분석하고, 문법적 관계를 파악합니다. 예를 들어, "나는 사과를 먹는다"라는 문장에서 주어(나), 목적어(사과), 동사(먹는다) 관계를 식별합니다.

1. **의미 분석 (Semantic Analysis)**: 문장이 전달하려는 의미를 추론합니다. 예를 들어, "배가 고프다"는 단순히 '배'라는 단어가 아니라 '배'가 '허기짐'을 의미하는지를 이해합니다.

1. **감정 분석 (Sentiment Analysis)**: 텍스트에 포함된 감정을 분석하여 긍정적, 부정적, 중립적 감정을 분류합니다. 예를 들어, "이 영화는 정말 재미있다"는 긍정적인 감정을, "이 영화는 지루하다"는 부정적인 감정을 나타냅니다.

1. **기계 번역 (Machine Translation)**: 한 언어에서 다른 언어로 번역하는 작업입니다. 예를 들어, 영어에서 한국어로 번역하는 시스템을 개발하는 것입니다.

1. **질의 응답 (Question Answering)**: 사용자가 입력한 질문에 대해 적절한 답을 제공하는 시스템입니다. 예를 들어, "파리의 수도는 어디인가요?"라는 질문에 대해 "파리는 프랑스의 수도입니다."라는 답을 제공하는 것입니다.

1. **텍스트 생성 (Text Generation)**: 주어진 주제나 조건에 맞춰 텍스트를 자동으로 생성하는 작업입니다. 예를 들어, 사용자가 제공한 주제에 대해 자동으로 기사를 작성하는 것이 이에 해당합니다.

자연어 처리는 다양한 분야에서 활용됩니다. 예를 들어, 검색 엔진, 음성 인식 시스템, 챗봇, 자동 번역 시스템 등에서 사용되며, 최근에는 대규모 언어 모델(예: GPT) 등의 발전으로 자연어 처리 기술이 더욱 진화하고 있습니다.


 # ■ 코퍼스 (corpus, 말뭉치)

자연어 분야의 연구를 위해 '특정한 목적'을 가지고 언어의 표본을 추출한 집합.

이를 사용하여 텍스트 데이터 분석및 저연어처리 (NLP) 에서 모델을 학습시키게 되는 데이터로 활용된다.

![](https://mblogthumb-phinf.pstatic.net/MjAyMjA1MDRfMTk4/MDAxNjUxNjI4NjQ4OTIx.ewwjMz43c_bhhXPLSv022-JqPZlMZmmjBh_P2_vSTfEg.y7dAYWl4kPg4Pm-30UIjKZoyiPriL6I0A9E8mjU83xQg.JPEG.mcstkorea/3._%EB%A7%90%EB%AD%89%EC%B9%98_%EC%A0%95%EC%9D%98.jpg?type=w800)


참조 사이트
국립국어원 말뭉치 검색
https://kcorpus.korean.go.kr/
![](https://img.sbs.co.kr/newsnet/test/sdf/20210927/1422164704/28394_1617005927.jpg)


# ■설치 패키지 & import


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import re

import tensorflow as tf
from tensorflow import keras

import torch
print('📍 torch version', torch.__version__)

# gensim: Generate Similar.
# 문서 간의 유사도(Similarity) 나 단어 의미 분석(Semantic Analysis) 에 초점을 둔
# 자연어 처리에 도움주는 패키지
!pip install gensim
import gensim

import nltk
nltk.download('punkt')  # 특정 데이터(리소스)를 컴퓨터로 다운로드
nltk.download('punkt_tab') # 'punkt_tab'의 역할: 문장 토큰화(Sentence Tokenization)를 수행하는 Punkt 알고리즘의 최신 리소스 파일입니다. 텍스트를 문장 단위로 쪼개거나 단어 단위로 분리할 때 "마침표(.)가 줄임말인지, 문장의 끝인지"를 판단하는 기준이 됩니다.
nltk.download('averaged_perceptron_tagger')

# konlpy : 한국여형태소분석기 (+품사태깅)
!pip install konlpy
import konlpy


📍 torch version 2.11.0+cpu
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.8/27.8 MB 25.7 MB/s eta 0:00:00


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.4/19.4 MB 22.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 438.5/438.5 kB 18.5 MB/s eta 0:00:00


# ■ 텍스트 전처리 (Text Preprocessing)

**텍스트 전처리**란?
주어진 문제를 해결하기 위해, 목적에 맞게 텍스트 데이터에 사전에 처리하는 작업입니다.

요리에 비유하자면, 아무리 조리방법과 장비가 좋다하더라도, 재료가 나쁘면, 혹은 재료가 제대로 손질되어 있지 않으면 좋은 요리가 나오기 힘듭니다.


텍스트 데이터 또한 적절한 전처리를 하지 않으면 앞으로 배울 자연어 처리 기법들이 제대로 동작하지 않습니다.
코퍼스 데이터는 사용하고자 하는 용도에 맞게 다양한 전처리 작업을 거치게 된다.

예를 들면.
- 토큰화(tokenization)
- 정제(cleaning)
- 정규화(normalization)



# ■ 02-01 토큰화(Tokenization)

- 주어진 코퍼스(corpus)에서 토큰(token)이라 불리는 단위로 나누는 작업을 토큰화(tokenization)라고 합니다

- 토큰의 단위는 상황에 따라 다르나, 보통 의미가 있는 단위로 토큰을 정의합니다.


## 🔹단어 토큰화(Word Tokenization)
토큰의 기준을 단어(word)로 하는 경우, 단어 토큰화(word tokenization)라고 합니다. 다만, 여기서 단어(word)는 단어 단위 외에도 단어구, 의미를 갖는 문자열로도 간주되기도 합니다.
예를 들어보겠습니다. 아래의 입력으로부터 구두점(punctuation)과 같은 문자는 제외시키는 간단한 단어 토큰화 작업을 해봅시다. 구두점이란 마침표(.), 컴마(,), 물음표(?), 세미콜론(;), 느낌표(!) 등과 같은 기호를 말합니다.


입력: **Time is an illusion. Lunchtime double so!**


이러한 입력으로부터 구두점을 제외시킨 토큰화 작업의 결과는 다음과 같습니다.


출력 : "Time", "is", "an", "illustion", "Lunchtime", "double", "so"


### NLTK 사용

In [2]:
sentence =  "Don't be fooled by the dark sounding name, Mr. Jone's Orphanage is as cheery as cheery goes for a pastry shop."

# Don't 어떻게 토큰화
# Don t
# Dont
# Do n't
# ...

In [3]:
# 영어 말뭉치(코퍼스)를 토큰화 하기 위한 도구들
from nltk.tokenize import word_tokenize
from nltk.tokenize import WordPunctTokenizer

In [4]:
print(word_tokenize(sentence))

['Do', "n't", 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', ',', 'Mr.', 'Jone', "'s", 'Orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop', '.']


In [5]:
# "Don't" -> "Do", "n't"
# "Jone's" -> "Jone", "s"

In [6]:
print(WordPunctTokenizer().tokenize(sentence))

['Don', "'", 't', 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', ',', 'Mr', '.', 'Jone', "'", 's', 'Orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop', '.']


### 📙TF/keras : text_to_word_sequence()


In [7]:
# 케라스 또한 토큰화 도구로서 text_to_word_sequence를 지원합니다.
#   텍스트를 소문자화(lowercase)
#   지정된 구두점 제거
#   공백 기준으로 단어 단위 토큰화

In [8]:
from tensorflow.keras.preprocessing.text import text_to_word_sequence

In [9]:
print(text_to_word_sequence(sentence))

["don't", 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', 'mr', "jone's", 'orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop']


### re 사용:
- pytorch 는 직접 대응되는 함수가 없지만 이와 같은 기능을 구현하는 방법들은 여럿 있다


In [10]:
import re

def tokenize_sequence(text, filters=r"[!\"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n]", lower=True, split=" "):
    if lower:
        text = text.lower()
    text = re.sub(filters, "", text)
    return [word for word in text.split(split) if word]

In [11]:
print(tokenize_sequence(sentence))

["don't", 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name,', 'mr.', "jone's", 'orphanage', 'is', 'as', 'cheery', 'as', 'cheery', 'goes', 'for', 'a', 'pastry', 'shop.']


### Hugging Face Tokenizers / Transformers 활용


In [12]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")   # 사전 학습 모델의 tokenizer 리턴.

print(tokenizer.tokenize(sentence))


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

['don', "'", 't', 'be', 'fooled', 'by', 'the', 'dark', 'sounding', 'name', ',', 'mr', '.', 'jon', '##e', "'", 's', 'orphanage', 'is', 'as', 'cheer', '##y', 'as', 'cheer', '##y', 'goes', 'for', 'a', 'pastry', 'shop', '.']


## 토큰화에서 고려해야할 사항

### 구두점과 특수문자
구두점과 특수문자를 무조건 제외하는게 답은 아니다


In [13]:
# 예]
#   단어 자체에 구두점 : m.p.h 나 Ph.D 나 AT&T
#   $, / 같은 특수문자 : $45.55 나 01/02/06
#   숫자 사이의 컴마: 123,456,789

### 띄어쓰기
줄임말과 단어 내에 띄어쓰기가 있는 경우.


In [14]:
# 예]
#  what're는 what are의 줄임말, we're는 we are의 줄임말

#  "New York" 이나 "rock 'n' roll" 는 띄어쓰기가 있지만 한 단어로 토큰화 할  필요도 있다.

### 표준 토큰화 예제

In [15]:
# Penn Treebank Tokenization의 규칙

# 규칙 1. 하이픈으로 구성된 단어는 하나로 유지한다.
# 규칙 2. doesn't와 같이 아포스트로피로 '접어'가 함께하는 단어는 분리해준다.


In [16]:
text = "Starting a home-based restaurant may be an ideal. it doesn't have a food chain or restaurant of their own."
text

"Starting a home-based restaurant may be an ideal. it doesn't have a food chain or restaurant of their own."

In [17]:
from nltk.tokenize import TreebankWordTokenizer

In [18]:
tokenizer = TreebankWordTokenizer()
print(tokenizer.tokenize(text))

['Starting', 'a', 'home-based', 'restaurant', 'may', 'be', 'an', 'ideal.', 'it', 'does', "n't", 'have', 'a', 'food', 'chain', 'or', 'restaurant', 'of', 'their', 'own', '.']


## 4. 🔹문장 토큰화(Sentence Tokenization)
- 토큰의 단위가 문장(sentence) 일 경우.


In [19]:
#  !, ? 는 문장구분을 위한 명확한 구분자 (boundary) 역할을 하지만
# 마침표(.) 는 그렇지 않다

In [20]:
# EX1)
# IP 192.168.56.31 서버에 들어가서 로그 파일 저장해서 aaa@gmail.com로 결과 좀 보내줘. 그 후 점심 먹으러 가자.

# EX2)
# Since I'm actively looking for Ph.D. students, I get the same question a dozen times every year.

### NLTK sent_tokenize()

In [21]:
from nltk.tokenize import sent_tokenize

In [22]:
text = "His barber kept his word. But keeping such a huge secret to himself was driving him crazy. Finally, the barber went up a mountain and almost to the edge of a cliff. He dug a hole in the midst of some reeds. He looked about, to make sure no one was near."
print('문장 토큰화1 :',sent_tokenize(text))


문장 토큰화1 : ['His barber kept his word.', 'But keeping such a huge secret to himself was driving him crazy.', 'Finally, the barber went up a mountain and almost to the edge of a cliff.', 'He dug a hole in the midst of some reeds.', 'He looked about, to make sure no one was near.']


In [23]:
text = "I am actively looking for Ph.D. students. and you are a Ph.D student."
print('문장 토큰화2 :',sent_tokenize(text))


문장 토큰화2 : ['I am actively looking for Ph.D. students.', 'and you are a Ph.D student.']


## 한국어 토큰화는 어렵다

In [24]:
# 교착어: 조사, 어미 등을 붙여서 말을 만드는 언어.


### 이유1) 한국어는 교착어이기 때문에

교착어의 의 특성
1. 조사
1. 띄어쓰기 단위
1. 형태소 (morpheme)

In [25]:
# 영어 '그' he/him 정도 뿐일텐데
#  한국어는  '그가', '그에게', '그를',  '그와',  '그는' ... 다양한 조사가 붙는다.

# ★ 대부분의 한국어 NLP 에선 조사는 분리해줄 필요가 있다.

# 한국어 토큰화 --> '형태소' 란 개념 반드시 이해
# 형태소(morpheme) : 뜻을 가진 가장 작은 말의 단위

In [26]:
# 문장 : "에디가 책을 읽었다"

# ▶이 문장을 띄어쓰기 단위 토큰화를 수행한다면
#   ['에디가', '책을', '읽었다']

# ▶ 형태소 단위로 분해하면
#   자립 형태소 : 에디, 책
#   의존 형태소 : -가, -을, 읽-, -었, -다

### 이유2) 한국어의 띄어쓰기는 영어보다 잘 지켜지지 않기 때문에


In [27]:
# EX1) 제가이렇게띄어쓰기를전혀하지않고글을썼다고하더라도글을이해할수있습니다.

# EX2) Tobeornottobethatisthequestion

## 품사 태깅 (Part-Of-Speech tagging, POS tagging)

In [28]:
# 'fly'  명사 - 파리
# 'fly'  동사 - 날다

# '못'   명사 - 못 박는다.
# '못'   부사 - 못한다.

# 단어 토큰화 과정에서 각 단어가 어떤 품사로 쓰였는지를 구분해놓기도 하는데,
# 이 작업을 품사 태깅(part-of-speech tagging)이라고 합니다.

## NLTK와 KoNLPY를 이용한 영어, 한국어 토큰화

In [29]:
from nltk.tag import pos_tag

In [30]:
text = "I am actively looking for Ph.D. students. and you are a Ph.D. student."
tokenized_sentence = word_tokenize(text)
print('단어 토큰화 :',tokenized_sentence)

단어 토큰화 : ['I', 'am', 'actively', 'looking', 'for', 'Ph.D.', 'students', '.', 'and', 'you', 'are', 'a', 'Ph.D.', 'student', '.']


In [31]:
nltk.download('averaged_perceptron_tagger_eng')
print('품사 태깅:', pos_tag(tokenized_sentence))

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger_eng.zip.


품사 태깅: [('I', 'PRP'), ('am', 'VBP'), ('actively', 'RB'), ('looking', 'VBG'), ('for', 'IN'), ('Ph.D.', 'NNP'), ('students', 'NNS'), ('.', '.'), ('and', 'CC'), ('you', 'PRP'), ('are', 'VBP'), ('a', 'DT'), ('Ph.D.', 'NNP'), ('student', 'NN'), ('.', '.')]


In [32]:
# Penn Treebank POG Tags에서 PRP는 인칭 대명사, VBP는 동사, RB는 부사, VBG는 현재부사,
# IN은 전치사, NNP는 고유 명사, NNS는 복수형 명사, CC는 접속사, DT는 관사를 의미합니다.

In [33]:
# KoNLPy  코엔엘파이 http://konlpy.org/ko/latest/

# 코엔엘파이를 통해서 사용할 수 있는 형태소 분석기로 Okt(Open Korea Text), 메캅(Mecab),
# 코모란(Komoran), 한나눔(Hannanum), 꼬꼬마(Kkma)가 있습니다.

In [34]:
from konlpy.tag import Okt
from konlpy.tag import Kkma

In [35]:
okt = Okt()
kkma = Kkma()

In [36]:
text = "오늘도 열심히 한 홍묵이, 내일도 화이팅"

In [37]:
okt.morphs(text)  # 형태소 분석

['오늘', '도', '열심히', '한', '홍묵', '이', ',', '내일', '도', '화이팅']

In [38]:
okt.pos(text)  # 품사 태깅

[('오늘', 'Noun'),
 ('도', 'Josa'),
 ('열심히', 'Adverb'),
 ('한', 'Verb'),
 ('홍묵', 'Noun'),
 ('이', 'Josa'),
 (',', 'Punctuation'),
 ('내일', 'Noun'),
 ('도', 'Josa'),
 ('화이팅', 'Noun')]

In [39]:
okt.nouns(text) # 명사만 추출

['오늘', '홍묵', '내일', '화이팅']

In [40]:
kkma.morphs(text)

['오늘', '도', '열심히', '한', '홍', '묵이', ',', '내일', '도', '화이', '팅']

In [41]:
kkma.pos(text)

[('오늘', 'NNG'),
 ('도', 'JX'),
 ('열심히', 'MAG'),
 ('한', 'MDN'),
 ('홍', 'NNG'),
 ('묵이', 'NNG'),
 (',', 'SP'),
 ('내일', 'NNG'),
 ('도', 'JX'),
 ('화이', 'XR'),
 ('팅', 'UN')]

In [42]:
kkma.nouns(text)

['오늘', '홍', '홍묵이', '묵이', '내일', '팅']

---
# ■ 02-02 정제(Cleaning), 정규화(Normalization)

- **정제(cleaning)** : 갖고 있는 코퍼스로부터 노이즈 데이터를 제거한다.
- **정규화(normalization)** : 표현 방법이 다른 단어들을 통합시켜서 같은 단어로 만들어준다.


## 규칙에 기반한 표기가 다른 단어들 통합

In [43]:
# ex)  USA 와 US,  uh-huh 와 uhhuh <= 형태는 다르지만 여전히 같은 의미

# 표기가 다른 단어들을 통합하는 방법중
#   - 어간추출 (stemming)
#   - 표제어추출(lemmatization)

## 대, 소문자 통합
영어권 언어의 또다른 정규화 방법

In [44]:
# 대소문자를 통합해야 하는 경우
# ex) Ferrari 로 검색했을때와 ferrari 로 검색했을때 동일한 결과를 얻을수 있게 해야 한다

# 대소문자가 구분시켜야 하는 경우
# ex) US 와 us 는 다른 의미의 단어.
# ex) 회사 이름(General Motors)나, 사람 이름(Bush) 등은 대문자로 유지

## 불용어(Stop word) 제거
불필요한 단어 제거

ex)
1. 등장빈도가 적은 단어
1. 길이가 짧은 단어



In [45]:
# 갖고 있는 데이터에서 유의미한 단어 토큰만을 선별하기 위해서는
# 큰 의미가 없는 단어 토큰을 제거해야 한다

# 예를 들면, I, my, me, over, 조사, 접미사 같은 단어들은 문장에서는 자주 등장하지만
# 실제 의미 분석을 하는데는 거의 기여하는 바가 없는 경우가 있습니다.
# 이러한 단어들을 불용어(stopword)라고 하며,
# NLTK에서는 위와 같은 100여개 이상의 영어 단어들을 불용어로 패키지 내에서 미리 정의하고 있습니다.

In [46]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [47]:
from nltk.corpus import stopwords

In [48]:
stop_words_list = stopwords.words('english')

In [49]:
len(stop_words_list)

198

In [50]:
print(stop_words_list[:10])

['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an']


### NLTK에서 불용어 제거하기

In [51]:
example = "Family is not an important thing. It's everything."
stop_words = set(stopwords.words('english'))

word_tokens = word_tokenize(example)

result = []

for word in word_tokens:
  if word not in stop_words:
    result.append(word)


print('불용어 제거 전 :',word_tokens)
print('불용어 제거 후 :',result)


불용어 제거 전 : ['Family', 'is', 'not', 'an', 'important', 'thing', '.', 'It', "'s", 'everything', '.']
불용어 제거 후 : ['Family', 'important', 'thing', '.', 'It', "'s", 'everything', '.']


### 한국어에서 불용어 제거하기

In [52]:
okt = Okt()

example = "고기를 아무렇게나 구우려고 하면 안 돼. 고기라고 다 같은 게 아니거든. 예컨대 삼겹살을 구울 때는 중요한 게 있지."
stop_words = "를 아무렇게나 구 우려 고 안 돼 같은 게 구울 때 는"

stop_words = set(stop_words.split())
word_tokens = okt.morphs(example)

result = [word for word in word_tokens if not word in stop_words]

print('불용어 제거 전 :',word_tokens)
print('불용어 제거 후 :',result)

불용어 제거 전 : ['고기', '를', '아무렇게나', '구', '우려', '고', '하면', '안', '돼', '.', '고기', '라고', '다', '같은', '게', '아니거든', '.', '예컨대', '삼겹살', '을', '구울', '때', '는', '중요한', '게', '있지', '.']
불용어 제거 후 : ['고기', '하면', '.', '고기', '라고', '다', '아니거든', '.', '예컨대', '삼겹살', '을', '중요한', '있지', '.']


In [53]:
# 아래의 링크는 보편적으로 선택할 수 있는 한국어 불용어 리스트를 보여줍니다.
# 하지만 여전히 절대적인 기준은 아닙니다.

# 링크 : https://www.ranks.nl/stopwords/korean

# 불용어가 많은 경우에는 코드 내에서 직접 정의하지 않고 txt 파일이나 csv 파일로 정리해놓고
# 이를 불러와서 사용하기도 합니다.


# ■ 02-06 정수 인코딩(Integer Encoding)
텍스트 -> 숫자

- 각 단어를 '고유한 정수' 에 맵핑(mapping) 시키는 전처리 작업


In [54]:
# 단어 인덱스 부여

# 일반적으로 단어의 등장 빈도수를 기준으로 정렬한뒤 부여함

## 정수 인코딩

### 1) dict 사용한 인코딩

In [55]:
from nltk.corpus import stopwords

In [56]:
raw_text = "A barber is a person. a barber is good person. a barber is huge person. he Knew A Secret! The Secret He Kept is huge secret. Huge secret. His barber kept his word. a barber kept his word. His barber kept his secret. But keeping and keeping such a huge secret to himself was driving the barber crazy. the barber went up a huge mountain."

In [57]:
sentences = sent_tokenize(raw_text)
sentences

['A barber is a person.',
 'a barber is good person.',
 'a barber is huge person.',
 'he Knew A Secret!',
 'The Secret He Kept is huge secret.',
 'Huge secret.',
 'His barber kept his word.',
 'a barber kept his word.',
 'His barber kept his secret.',
 'But keeping and keeping such a huge secret to himself was driving the barber crazy.',
 'the barber went up a huge mountain.']

In [58]:
# 정제, 정규화, 단어 토큰화, 단어 빈도

vocab = {}  # 각 단어별 빈도수
preprocessed_sentences = []
stop_words = set(stopwords.words('english'))

for sentence in sentences:
  result = []  # 전처리된 문장의 단어들

  # 각 문장마다 단어 토큰화
  tokenized_sentence = word_tokenize(sentence)

  for word in tokenized_sentence:  # 각 토큰마다
    word = word.lower()  # 모든 단어를 소문자화 단어의 개수를 줄인다.
    if word not in stop_words:  # 불용어 제거
      if len(word) > 2:  # 단어 길이가 2이하인 경우에 대하여 추가로 단어 제거
        result.append(word)

        # 빈도수 체크
        if word not in vocab:
          vocab[word] = 0
        vocab[word] += 1

  preprocessed_sentences.append(result)

preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [59]:
# 단어 빈도수
vocab

{'barber': 8,
 'person': 3,
 'good': 1,
 'huge': 5,
 'knew': 1,
 'secret': 6,
 'kept': 4,
 'word': 2,
 'keeping': 2,
 'driving': 1,
 'crazy': 1,
 'went': 1,
 'mountain': 1}

In [60]:
vocab['barber']

8

In [61]:
# 빈도수가 높은 순서대로 정렬
vocab_sorted = sorted(vocab.items(), key = lambda x:x[1], reverse=True)
vocab_sorted

[('barber', 8),
 ('secret', 6),
 ('huge', 5),
 ('kept', 4),
 ('person', 3),
 ('word', 2),
 ('keeping', 2),
 ('good', 1),
 ('knew', 1),
 ('driving', 1),
 ('crazy', 1),
 ('went', 1),
 ('mountain', 1)]

In [62]:
# 높은 빈도수를 가진 단어에 낮은 정수 인덱스 부여.  정수값은 1부터 부여

word_to_index= {}

idx = 0
for word, freq in vocab_sorted:
  if freq > 1:  # 빈도수가 작은 단어는 정수 인덱싱 제외.
    idx += 1
    word_to_index[word] = idx

word_to_index

{'barber': 1,
 'secret': 2,
 'huge': 3,
 'kept': 4,
 'person': 5,
 'word': 6,
 'keeping': 7}

In [63]:
# 위에서 자주 사용하는 단어 (어휘) 5개까지만 사용하려면

vocab_size = 5

word_to_index = dict(list(word_to_index.items())[:vocab_size])

word_to_index

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5}

In [64]:
#  'But keeping and keeping such a huge secret to himself was driving the barber crazy.',

#   ↓  정제 & 토큰화

#  ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],

#   ↓  정수인덱싱 하게 되면?

#  [??,  ??,  3,  2,  ??,  1,  ??]

#  ↑ OOV  어휘 사전에 없는 단어들이 존재할것이다!

### OOV (Out-Of-Vocabulary)

In [65]:
# OOV 를 word_to_index  에 새롭게 추가
# 단어집합(사전) 에 없는 단어는 'OOV' 로 인덱싱 할거다.

# ['barber', 'person'] -> [1, 5]
# ['barber', 'good', 'person'] -> [1, ??, 5]   <- 단어집합에 없는 단어는 OOV 로 인코딩

In [66]:
# OOV 를 사전에 추가
word_to_index['OOV'] = len(word_to_index) + 1

word_to_index

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5, 'OOV': 6}

In [67]:
# word_to_index 를 사용하여 sentences 의 모든 단어들을 정수로 인코딩

encoded_sentences = []

for sentence in preprocessed_sentences:
  encoded_sentence = []

  for word in sentence:
    try:
      encoded_sentence.append(word_to_index[word])
    except KeyError:
      encoded_sentence.append(word_to_index['OOV'])

  encoded_sentences.append(encoded_sentence)

encoded_sentences

[[1, 5],
 [1, 6, 5],
 [1, 3, 5],
 [6, 2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4, 6],
 [1, 4, 6],
 [1, 4, 2],
 [6, 6, 3, 2, 6, 1, 6],
 [1, 6, 3, 6]]

In [68]:
# 원본 문장과 인코딩된 문장(정수 시퀀스)을 함 같이 출력해보자
for e, s in zip(encoded_sentences, sentences):
  print(f'{str(e):25} {s}')

[1, 5]                    A barber is a person.
[1, 6, 5]                 a barber is good person.
[1, 3, 5]                 a barber is huge person.
[6, 2]                    he Knew A Secret!
[2, 4, 3, 2]              The Secret He Kept is huge secret.
[3, 2]                    Huge secret.
[1, 4, 6]                 His barber kept his word.
[1, 4, 6]                 a barber kept his word.
[1, 4, 2]                 His barber kept his secret.
[6, 6, 3, 2, 6, 1, 6]     But keeping and keeping such a huge secret to himself was driving the barber crazy.
[1, 6, 3, 6]              the barber went up a huge mountain.


## 2. 📙케라스(Keras)의 텍스트 전처리


In [7]:
from tensorflow.keras.preprocessing.text import Tokenizer

### Tokenizer

**tf.keras.preprocessing.text.Tokenizer**
- fit_on_texts()
- word_index
- word_counts
- texts_to_sequences()
- num_words =
- oov_token =

---

https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/text/Tokenizer

```python
tf.keras.preprocessing.text.Tokenizer(
    num_words=None,
    filters=&#x27;!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n',
    lower=True,
    split=&#x27; ',
    char_level=False,
    oov_token=None,
    analyzer=None,
    **kwargs
)
```


In [70]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

### fit_on_texts()

In [71]:
tokenizer = Tokenizer()

In [72]:
# fit_on_texts()안에 코퍼스를 입력으로 하면 빈도수를 기준으로 단어 집합을 생성.
tokenizer.fit_on_texts(preprocessed_sentences)

### `.word_index`, `.word_counts`

In [73]:
print(tokenizer.word_index)  # 사전 정보

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5, 'word': 6, 'keeping': 7, 'good': 8, 'knew': 9, 'driving': 10, 'crazy': 11, 'went': 12, 'mountain': 13}


In [74]:
print(tokenizer.word_counts)  # 등장 빈도

OrderedDict({'barber': 8, 'person': 3, 'good': 1, 'huge': 5, 'knew': 1, 'secret': 6, 'kept': 4, 'word': 2, 'keeping': 2, 'driving': 1, 'crazy': 1, 'went': 1, 'mountain': 1})


### `text_to_sequence()`

In [75]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [76]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[1, 5],
 [1, 8, 5],
 [1, 3, 5],
 [9, 2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4, 6],
 [1, 4, 6],
 [1, 4, 2],
 [7, 7, 3, 2, 10, 1, 11],
 [1, 12, 3, 13]]

### `num_words=` 단어 사전의 크기

In [77]:
# Tokenizer (num_words=)
#   빈도수가 높은 상위 몇개의 단어만 사용 가능.

In [78]:
vocab_size = 5
tokenizer = Tokenizer(num_words=vocab_size + 1)  # 빈도수 상위 5개 단어만 사용
tokenizer.fit_on_texts(preprocessed_sentences)

In [79]:
# num_words 는 숫자를 0부터 카운트
# 만약 숫자 5를 설정하면 0 ~ 4번 단어 보존. => 1번단어 ~ 4번단어 (4개 단어)만 남는다.

# 그래서 1번 ~ 5번 단어(5개)를 사용하고 싶다면 5 + 1 값을 넣어준겁니다.

# keras Tokenizer 는 0번을 자연어처리에서 필요한 패딩(padding) 에 사용함.

In [80]:
print(tokenizer.word_index)

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5, 'word': 6, 'keeping': 7, 'good': 8, 'knew': 9, 'driving': 10, 'crazy': 11, 'went': 12, 'mountain': 13}


In [81]:
# 상위 5개만 사용하는 옵션을 줬는데 여전히 모든 단어가 출력?

# ↑ 일단 모든 단어에 대한 인덱스 정보는 담겨 있다
# ↓ .word_counts 도 마찬가지

In [82]:
print(tokenizer.word_counts)

OrderedDict({'barber': 8, 'person': 3, 'good': 1, 'huge': 5, 'knew': 1, 'secret': 6, 'kept': 4, 'word': 2, 'keeping': 2, 'driving': 1, 'crazy': 1, 'went': 1, 'mountain': 1})


In [83]:
# num_words= 가 적용되는 것은 texts_to_sequences() 를 사용할때 적용됨!

In [84]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [85]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[1, 5],
 [1, 5],
 [1, 3, 5],
 [2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4],
 [1, 4],
 [1, 4, 2],
 [3, 2, 1],
 [1, 3]]

In [86]:
# ↑상위 5개 단어만 (1번~5번) 보존되고 나머지 단어들은 제거된 것과 같은 결과.

"""
[[1, 5],      <-   ['barber', 'person'],
 [1, 5],      <-   ['barber', 'good', 'person']  num_words= 바깥의 단어 'good' 는 제거된 결과
 [1, 3, 5],
 [2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4],
 [1, 4],
 [1, 4, 2],
 [3, 2, 1],
 [1, 3]]
코퍼스에 대해서 각 단어를 이미 정해진 인덱스로 변환하는데,
상위 5개의 단어만을 사용하겠다고 지정하였으므로 1번 단어부터 5번 단어까지만 보존되고
num_words= 바깥의 단어 단어들은 제거된 것을 볼 수 있습니다.
"""
None

# 만약 word_index 와 word counts 에서도
# 지정된 num_words= 만큼의 단어만 남기고 싶다면! ↓↓

In [87]:
tokenizer = Tokenizer()  # num_words 없음
tokenizer.fit_on_texts(preprocessed_sentences)

In [88]:
vocab_size = 5

tokenizer.word_index = dict(list(tokenizer.word_index.items())[:vocab_size])

tokenizer.word_counts = dict(sorted(tokenizer.word_counts.items(), key=lambda x: x[1], reverse=True))

In [89]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[1, 5],
 [1, 5],
 [1, 3, 5],
 [2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4],
 [1, 4],
 [1, 4, 2],
 [3, 2, 1],
 [1, 3]]

In [90]:
tokenizer.word_index

{'barber': 1, 'secret': 2, 'huge': 3, 'kept': 4, 'person': 5}

#### oov_token=
- 단어집합에 없는 단어들 (out of vocabulary)
- keras 는 'OOV' 인덱스를 1 로 설정 (기본)


In [91]:
# 숫자 0 (패딩용) 과 OOV를 고려해서 단어 집합의 크기는 +2
# 실제 단어는 인덱스 2부터 시작
vocab_size = 5

tokenizer = Tokenizer(num_words=vocab_size + 2, oov_token='OOV')
tokenizer.fit_on_texts(preprocessed_sentences)

In [92]:
tokenizer.word_index

{'OOV': 1,
 'barber': 2,
 'secret': 3,
 'huge': 4,
 'kept': 5,
 'person': 6,
 'word': 7,
 'keeping': 8,
 'good': 9,
 'knew': 10,
 'driving': 11,
 'crazy': 12,
 'went': 13,
 'mountain': 14}

In [93]:
tokenizer.word_index['OOV']

1

In [94]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [95]:
tokenizer.texts_to_sequences(preprocessed_sentences)

[[2, 6],
 [2, 1, 6],
 [2, 4, 6],
 [1, 3],
 [3, 5, 4, 3],
 [4, 3],
 [2, 5, 1],
 [2, 5, 1],
 [2, 5, 3],
 [1, 1, 4, 3, 1, 2, 1],
 [2, 1, 4, 1]]

In [96]:
# ↑  빈도수 상위 5개 단어는   인덱스 2 ~ 6까지로 인코딩.
# 그 외의 단어들은 'OOV' 인덱스 1 로 인코딩

## HuggingFace Tokenizers

In [97]:
# BERT, GPT, RoBERTa 같은 '사전학습된 모델'을 쓴다면,
# Hugging Face의 transformers 라이브러리가 더 강력하고 표준적인 선택입니다.

# 장점!
#  - WordPiece/BPE 토크나이징 자동 지원
#  - 패딩, 마스크, 어텐션 생성까지 자동
#  - PyTorch 텐서 바로 반환
# ⭐️ Keras보다 훨씬 강력하고 범용적인 방법입니다.

In [98]:
from transformers import AutoTokenizer

In [99]:
# AutoTokenizer는 모델 이름만 지정하면 해당 모델에 맞는 토크나이저 클래스를 자동으로 로드해 줍니다
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

text = ["I love Pytorch", "It is great"]

encodings = tokenizer(text, padding=True, truncatuin=True, return_tensors="pt")

encodings

{'input_ids': tensor([[  101,  1045,  2293,  1052, 22123,  2953,  2818,   102],
        [  101,  2009,  2003,  2307,   102,     0,     0,     0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 0, 0, 0]])}

In [100]:
encodings.input_ids

# 토큰의 정수 인덱스.

# 텍스트를 모델의 어휘(vocabulary) 에 따라 숫자로 바꾼 결과.
# 즉, 각 단어(혹은 서브워드, subword)에 해당하는 정수 인덱스 리스트

tensor([[  101,  1045,  2293,  1052, 22123,  2953,  2818,   102],
        [  101,  2009,  2003,  2307,   102,     0,     0,     0]])

In [101]:
encodings.token_type_ids

# BERT 같은 모델은 두 문장 입력(문장쌍) 을 처리할 수 있는데,
# 이때 “이 토큰이 어느 문장에 속하는가”를 알려주는 역할을 합니다.

# 첫 번째 문장: 0
# 두 번째 문장: 1

tensor([[0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0]])

In [102]:
encodings.attention_mask

# 패딩 무시용 마스크

# 모델이 “어떤 토큰이 실제 단어이고, 어떤 부분이 단순히 패딩(padding)”인지를 구분하도록 도와줍니다.
# 실제 단어 위치 → 1
# 패딩된 위치 → 0

tensor([[1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 0, 0, 0]])

## 한국어 사전학습 모델 사용

사용할수 있는 한국어 모델들 (2025.7 현재)

| 모델 이름                                      | 설명                  |
| ------------------------------------------ | ------------------- |
| `klue/bert-base`                           | KLUE 벤치마크용 BERT     |
| `beomi/KcELECTRA-base`                     | SNS 대화체 한국어 ELECTRA |
| `snunlp/KR-ELECTRA-discriminator`          | 서울대 NLP팀 ELECTRA    |
| `monologg/koelectra-base-v3-discriminator` | monologg의 KoELECTRA |
| `kykim/bert-kor-base`                      | 표준 한국어 BERT         |
| `jhgan/ko-sbert-multitask`                 | 문장 임베딩용 SBERT       |


In [103]:
# KLUE 한국어 BERT 모델: https://huggingface.co/klue/bert-base
# KLUE(Korean Language Understanding Evaluation)라는 한국어 NLP 벤치마크 프로젝트를 만든 팀
#    (네이버, 업스테이지 등 협업)의 HuggingFace 조직 계정명
# ※ klue/bert-base <- 조직(또는 사용자)명 / 모델명 형식의 HuggingFace 네이밍 규칙

text = "안녕하세요. 오늘 날씨가 정말 좋네요!"

tokenizer = AutoTokenizer.from_pretrained('klue/bert-base')
tokens = tokenizer.tokenize(text)

print(tokens)

config.json:   0%|          | 0.00/425 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/289 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/248k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/495k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

['안녕', '##하', '##세요', '.', '오늘', '날씨', '##가', '정말', '좋', '##네', '##요', '!']


In [104]:
# KoBERT
!pip install kobert-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 36.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 20.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 30.0 MB/s eta 0:00:00
  Attempting uninstall: huggingface-hub
    Found existing installation: huggingface_hub 1.28.0
    Uninstalling huggingface_hub-1.28.0:
      Successfully uninstalled huggingface_hub-1.28.0
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.23.1
    Uninstalling tokenizers-0.23.1:
      Successfully uninstalled tokenizers-0.23.1
  Attempting uninstall: transformers
    Found existing installation: transformers 5.16.1
    Uninstalling transformers-5.16.1:
      Successfully uninstalled transformers-5.16.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the

In [105]:
from kobert_transformers import get_tokenizer

tokenizer = get_tokenizer()
print(tokenizer.tokenize('한국어 토크나이저 테스트'))

ImportError: cannot import name 'download_url' from 'transformers.utils' (/usr/local/lib/python3.13/dist-packages/transformers/utils/__init__.py)

### special tokens
`AutoTokenizer` (예: BERT 계열 등)에는 **모델 내부적으로 특별한 의미를 가지는 “스페셜 토큰(special tokens)”** 이 여러 개 정의되어 있다.
이 토큰들은 단순한 단어가 아니라, **모델이 문장의 구조를 이해하기 위해 꼭 필요한 신호 역할**을 한다.

---

## 🧩 대표적인 Special Tokens (BERT 기준)

| 토큰                          | 예시 ID (bert-base-uncased 기준) | 의미              | 역할                                |
| --------------------------- | ---------------------------- | --------------- | --------------------------------- |
| `[PAD]`                     | `0`                          | Padding Token   | 짧은 문장 길이 맞추기용                     |
| `[UNK]`                     | `100`                        | Unknown Token   | 어휘(vocab)에 없는 단어                  |
| `[CLS]`                     | `101`                        | Class Token     | 문장 전체 표현 (classification용)        |
| `[SEP]`                     | `102`                        | Separator Token | 문장 구분 (문장 A/B 구분)                 |
| `[MASK]`                    | `103`                        | Mask Token      | Masked Language Modeling에서 가려진 단어 |
| `[unused0]` ~ `[unused999]` | (104~)                       | 미사용 토큰          | 예비 슬롯 (특수 태그용으로 종종 활용)            |

> 💡 **ID 값은 모델마다 다를 수 있습니다!**
> 예: RoBERTa나 DistilBERT는 `[CLS]`가 `<s>`로, `[SEP]`가 `</s>`로 이름이 다릅니다.

---

## 🔍 각 토큰의 구체적 역할

### 1️⃣ `[CLS]` (Classify)

* 문장 맨 앞에 항상 추가됨
* 문장 전체의 의미를 요약하는 “대표 토큰”
* 분류(Classification) 작업 시 이 토큰의 벡터를 사용

**예시:**

```python
[CLS] I love PyTorch [SEP]
```

---

### 2️⃣ `[SEP]` (Separator)

* 문장 끝 혹은 문장 사이에 추가됨
* 문장 경계를 표시하거나 문장쌍(A/B) 구분용으로 사용

**예시:**

```python
[CLS] I love PyTorch [SEP] It is great [SEP]
```

---

### 3️⃣ `[PAD]`

* 배치 입력 시 문장 길이를 맞추기 위해 뒤쪽에 채워짐
* 모델이 무시하도록 `attention_mask`에서 0으로 표시됨

```python
[CLS] I love PyTorch [SEP] [PAD] [PAD]
```

---

### 4️⃣ `[UNK]`

* 토크나이저의 어휘 사전에 없는 단어가 들어왔을 때 사용

```python
"I love 🦖" → [CLS] I love [UNK] [SEP]
```

---

### 5️⃣ `[MASK]`

* 사전학습(Pretraining) 중 일부 단어를 가려서 모델이 예측하게 만듦
* Masked Language Modeling(MLM)에서 사용

```python
[CLS] I love [MASK] [SEP]
```

모델은 `[MASK]` 위치의 단어를 예측합니다 → `"PyTorch"`

---


## 📚 모델별 Special Token 비교

| 모델                          | 문장 시작   | 문장 종료   | 패딩      | 마스크              | 비고                 |
| --------------------------- | ------- | ------- | ------- | ---------------- | ------------------ |
| **BERT**                    | `[CLS]` | `[SEP]` | `[PAD]` | `[MASK]`         | 전통적인 구조            |
| **RoBERTa**                 | `<s>`   | `</s>`  | `<pad>` | `<mask>`         | BERT 변형            |
| **DistilBERT**              | `[CLS]` | `[SEP]` | `[PAD]` | `[MASK]`         | BERT 경량화           |
| **GPT-2 / GPT-3 / GPT-Neo** | 없음      | ``      | 없음      | 없음               | 단방향 언어모델           |
| **T5**                      | `<pad>` | `</s>`  | `<pad>` | `<extra_id_0>` 등 | Encoder-Decoder 구조 |

---

## 🧠 요약

| 토큰       | 용도           | 일반적 ID (BERT 기준) |
| -------- | ------------ | ---------------- |
| `[PAD]`  | 패딩           | `0`              |
| `[CLS]`  | 문장 대표        | `101`            |
| `[SEP]`  | 문장 구분        | `102`            |
| `[MASK]` | 단어 가리기 (MLM) | `103`            |
| `[UNK]`  | 미등록 단어       | `100`            |



In [106]:
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')
print(tokenizer.special_tokens_map)

{'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}


# ■ 패딩 (Padding)

In [ ]:
# 자연어 처리를 하다보면 각 문장(또는 문서)은 서로 길이가 다를 수 있습니다.
# 그런데 기계는 길이가 전부 동일한 문서들에 대해서는 하나의 행렬로 보고,
# 한꺼번에 묶어서 처리할 수 있습니다.

# 다시 말해 병렬 연산을 위해서 여러 문장의 길이를 임의로 동일하게 맞춰주는 작업이 필요!.

![](https://miro.medium.com/v2/resize:fit:1400/1*CPLhZoVSTCWgAxe2LKXoOA.png)

In [107]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [108]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(preprocessed_sentences)

encoded = tokenizer.texts_to_sequences(preprocessed_sentences)
print(encoded)

[[1, 5], [1, 8, 5], [1, 3, 5], [9, 2], [2, 4, 3, 2], [3, 2], [1, 4, 6], [1, 4, 6], [1, 4, 2], [7, 7, 3, 2, 10, 1, 11], [1, 12, 3, 13]]


In [109]:
# 모두 동일한 길이로 맞춰주기 위해서 이 중에서 가장 길이가 긴 문장의 길이를 계산
max_len = max(len(item) for item in encoded)
print(max_len)

7


In [110]:
# 모든 문장의 길이를 max_len 으로 맞추어봅니다
# 패딩문자는 'PAD' 라 하고, 정수 0 으로 정의
# 길이가 7보다 짧은 문장에는 숫자 0 으로 채워짐.

In [111]:
# 수동으로 작성
print(encoded)

for sentence in encoded:
  while len(sentence) < max_len:
    sentence.append(0)

print(encoded)

[[1, 5], [1, 8, 5], [1, 3, 5], [9, 2], [2, 4, 3, 2], [3, 2], [1, 4, 6], [1, 4, 6], [1, 4, 2], [7, 7, 3, 2, 10, 1, 11], [1, 12, 3, 13]]
[[1, 5, 0, 0, 0, 0, 0], [1, 8, 5, 0, 0, 0, 0], [1, 3, 5, 0, 0, 0, 0], [9, 2, 0, 0, 0, 0, 0], [2, 4, 3, 2, 0, 0, 0], [3, 2, 0, 0, 0, 0, 0], [1, 4, 6, 0, 0, 0, 0], [1, 4, 6, 0, 0, 0, 0], [1, 4, 2, 0, 0, 0, 0], [7, 7, 3, 2, 10, 1, 11], [1, 12, 3, 13, 0, 0, 0]]


## keras 로 패딩 : `pad_sequences`

In [112]:
from keras.utils import pad_sequences

https://www.tensorflow.org/api_docs/python/tf/keras/utils/pad_sequences
```python
tf.keras.utils.pad_sequences(
    sequences,
    maxlen=None,
    dtype='int32',
    padding='pre',  # pre: 시퀀스 '앞' 에 패딩.   post: 시퀀스 '뒤'에 패딩
    truncating='pre', #sequence 가 maxlen보다 긴 경우 잘라내기.
    value=0.0  # 패딩값.
)
```

In [113]:
preprocessed_sentences

[['barber', 'person'],
 ['barber', 'good', 'person'],
 ['barber', 'huge', 'person'],
 ['knew', 'secret'],
 ['secret', 'kept', 'huge', 'secret'],
 ['huge', 'secret'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'word'],
 ['barber', 'kept', 'secret'],
 ['keeping', 'keeping', 'huge', 'secret', 'driving', 'barber', 'crazy'],
 ['barber', 'went', 'huge', 'mountain']]

In [114]:
encoded = tokenizer.texts_to_sequences(preprocessed_sentences)
encoded

[[1, 5],
 [1, 8, 5],
 [1, 3, 5],
 [9, 2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4, 6],
 [1, 4, 6],
 [1, 4, 2],
 [7, 7, 3, 2, 10, 1, 11],
 [1, 12, 3, 13]]

In [116]:
padded = pad_sequences(encoded)
padded

array([[ 0,  0,  0,  0,  0,  1,  5],
       [ 0,  0,  0,  0,  1,  8,  5],
       [ 0,  0,  0,  0,  1,  3,  5],
       [ 0,  0,  0,  0,  0,  9,  2],
       [ 0,  0,  0,  2,  4,  3,  2],
       [ 0,  0,  0,  0,  0,  3,  2],
       [ 0,  0,  0,  0,  1,  4,  6],
       [ 0,  0,  0,  0,  1,  4,  6],
       [ 0,  0,  0,  0,  1,  4,  2],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 0,  0,  0,  1, 12,  3, 13]], dtype=int32)

In [117]:
padded = pad_sequences(encoded, padding='post')
padded

array([[ 1,  5,  0,  0,  0,  0,  0],
       [ 1,  8,  5,  0,  0,  0,  0],
       [ 1,  3,  5,  0,  0,  0,  0],
       [ 9,  2,  0,  0,  0,  0,  0],
       [ 2,  4,  3,  2,  0,  0,  0],
       [ 3,  2,  0,  0,  0,  0,  0],
       [ 1,  4,  6,  0,  0,  0,  0],
       [ 1,  4,  6,  0,  0,  0,  0],
       [ 1,  4,  2,  0,  0,  0,  0],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 1, 12,  3, 13,  0,  0,  0]], dtype=int32)

In [118]:
padded = pad_sequences(encoded, padding='post', maxlen=5)
padded

array([[ 1,  5,  0,  0,  0],
       [ 1,  8,  5,  0,  0],
       [ 1,  3,  5,  0,  0],
       [ 9,  2,  0,  0,  0],
       [ 2,  4,  3,  2,  0],
       [ 3,  2,  0,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  2,  0,  0],
       [ 3,  2, 10,  1, 11],
       [ 1, 12,  3, 13,  0]], dtype=int32)

In [119]:
encoded

[[1, 5],
 [1, 8, 5],
 [1, 3, 5],
 [9, 2],
 [2, 4, 3, 2],
 [3, 2],
 [1, 4, 6],
 [1, 4, 6],
 [1, 4, 2],
 [7, 7, 3, 2, 10, 1, 11],
 [1, 12, 3, 13]]

In [120]:
padded = pad_sequences(encoded, padding='post', maxlen=5, truncating='post')
padded

array([[ 1,  5,  0,  0,  0],
       [ 1,  8,  5,  0,  0],
       [ 1,  3,  5,  0,  0],
       [ 9,  2,  0,  0,  0],
       [ 2,  4,  3,  2,  0],
       [ 3,  2,  0,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  6,  0,  0],
       [ 1,  4,  2,  0,  0],
       [ 7,  7,  3,  2, 10],
       [ 1, 12,  3, 13,  0]], dtype=int32)

In [121]:
# [ 7,  7,  3,  2, 10,  1, 11]  <- 원본 문장
#  [ 3,  2, 10,  1, 11]  <==  truncating='pre'
#  [ 7,  7,  3,  2, 10]   <-- truncating='post'

### value=
패딩 문자 값 지정

In [122]:
# value=
# 패딩 문자 값 지정

# 패딩값은 일반적으로는 '0' 값을 사용함.

# 다른 숫자로도 패딩할수 있다.
# ex) 단어집합(사전)의 크기에 +1 한 숫자를 패딩문자로 사용한다면..

In [123]:
last_value = len(tokenizer.word_index) + 1
last_value

14

In [124]:
padded = pad_sequences(encoded, padding='post', value=last_value)
padded

array([[ 1,  5, 14, 14, 14, 14, 14],
       [ 1,  8,  5, 14, 14, 14, 14],
       [ 1,  3,  5, 14, 14, 14, 14],
       [ 9,  2, 14, 14, 14, 14, 14],
       [ 2,  4,  3,  2, 14, 14, 14],
       [ 3,  2, 14, 14, 14, 14, 14],
       [ 1,  4,  6, 14, 14, 14, 14],
       [ 1,  4,  6, 14, 14, 14, 14],
       [ 1,  4,  2, 14, 14, 14, 14],
       [ 7,  7,  3,  2, 10,  1, 11],
       [ 1, 12,  3, 13, 14, 14, 14]], dtype=int32)

## Huggingface Tokenizers

In [ ]:
# transformers.AutoTokenizer의 padding= 과 truncation= 옵션은
# 토크나이징 시 입력 길이를 맞추기 위해 사용됨.
# 이 두 가지는 특히 배치(batch) 입력 시 매우 중요

###  padding=
문장 길이 맞추기 (짧은 문장 뒤에 0을 채움)

In [ ]:
# padding=
# 문장 길이 맞추기 (짧은 문장 뒤에 0을 채움)

# 모든 문장이 길이가 다르기 때문에, 배치 텐서로 묶기 전에 길이를 통일해야 합니다.
# 짧은 문장에는 [PAD] 토큰(보통 id=0)을 추가해서 길이를 맞춥니다.

In [130]:
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")

batch = tokenizer(
    ["I love PyTorch", "PyTorch is great and lovely"],
    padding='max_length',  # 패딩 적용
    return_tensors="pt",
    max_length=5,
    truncation=True,
)

print(batch["input_ids"])

tensor([[  101,  1045,  2293,  1052,   102],
        [  101,  1052, 22123,  2953,   102]])


# ■ 02-08 원-핫 인코딩(One-Hot Encoding)

![](https://heung-bae-lee.github.io/image/why_is_sparse_matirx_one_hot_encoding.png)


## 단어집합 (vocabulary)

In [ ]:
# 원-핫 인코딩을 위해서 먼저 해야할 일은 '단어 집합'을 만드는 일입니다.
# 텍스트의 모든 단어를 '중복을 허용하지 않고' 모아놓으면 이를 단어 집합

# 그리고 이 단어 집합에 고유한 정수를 부여하는 정수 인코딩을 진행합니다.
# 텍스트에 단어가 총 5,000개가 존재한다면, 단어 집합의 크기는 5,000입니다.

# 각 단어에 고유한 정수 인덱싱이 되었다고 할때
# 이 '숫자'로 바뀐 단어들을 '벡터' 로 다루고 싶다면?

## 원핫 인코딩(One-hot encoding) 작성(수동)

In [131]:
# 단어집합의 크기를 벡터의 '차원' 으로 하고
# 표현하고 싶은 단어의 인덱스에 '1' 값을 부여, 다른 인덱스에는 '0' 을 부여하여 단어의 벡터 표현
# 이렇게 표현된 벡터를 원-핫 벡터 (one-hot vector)

In [132]:
tokens = okt.morphs("나는 자연어 처리를 배운다")
print(tokens)

['나', '는', '자연어', '처리', '를', '배운다']


In [133]:
# 각 토큰에 대해서 고유한 정수를 부여합니다.
# 지금은 문장이 짧기 때문에 각 단어의 빈도수를 고려하지 않지만,
# 빈도수 순으로 단어를 정렬하여 정수를 부여하는 경우가 많습니다.

word_to_index = {word: index for index, word in enumerate(tokens)}
print(word_to_index)


{'나': 0, '는': 1, '자연어': 2, '처리': 3, '를': 4, '배운다': 5}


In [134]:
def one_hot_encoding(word,word_to_index):
  one_hot_vector = [0] * len(word_to_index)  # 단어 집합의 크기를 벡터의 '차원'으로
  index = word_to_index[word]
  one_hot_vector[index] = 1  # 포현하고 싶은 단어의 인데스에 '1'값 부여

  return one_hot_vector

In [135]:
one_hot_encoding('자연어', word_to_index)

[0, 0, 1, 0, 0, 0]

## keras 의 원-핫 인코딩 `to_categorical()`

In [136]:
from keras.utils import to_categorical

https://www.tensorflow.org/api_docs/python/tf/keras/utils/to_categorical
```python
tf.keras.utils.to_categorical(
    x, num_classes=None
)
```


In [138]:
text = "나랑 점심 먹으러 갈래 점심 메뉴는 햄버거 갈래 갈래 햄버거 최고야"
# 위와 같은 문장이 있다고 했을 때, 케라스 토크나이저를 이용한 정수 인코딩은 다음과 같습니다.
tokenizer = Tokenizer()
tokenizer.fit_on_texts([text])
print('단어 집합 :',tokenizer.word_index)

단어 집합 : {'갈래': 1, '점심': 2, '햄버거': 3, '나랑': 4, '먹으러': 5, '메뉴는': 6, '최고야': 7}


In [139]:
sub_text = "점심 먹으러 갈래 메뉴는 햄버거 최고야"
encoded = tokenizer.texts_to_sequences([sub_text])[0]
print(encoded)

[2, 5, 1, 6, 3, 7]


In [140]:
# 위 정수 인코딩 된 결과 (6,) 를 원-핫 인코딩 해보자. (6, 8)

one_hot = to_categorical(encoded)
one_hot

array([[0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1.]])

In [141]:
"""
[[0. 0. 1. 0. 0. 0. 0. 0.] # 인덱스 2의 원-핫 벡터
 [0. 0. 0. 0. 0. 1. 0. 0.] # 인덱스 5의 원-핫 벡터
 [0. 1. 0. 0. 0. 0. 0. 0.] # 인덱스 1의 원-핫 벡터
 [0. 0. 0. 0. 0. 0. 1. 0.] # 인덱스 6의 원-핫 벡터
 [0. 0. 0. 1. 0. 0. 0. 0.] # 인덱스 3의 원-핫 벡터
 [0. 0. 0. 0. 0. 0. 0. 1.]] # 인덱스 7의 원-핫 벡터

위의 결과는 "점심 먹으러 갈래 메뉴는 햄버거 최고야"라는 문장이
[2, 5, 1, 6, 3, 7]로 정수 인코딩이 되고나서,
각각의 인코딩 된 결과를 인덱스로 원-핫 인코딩이 수행된 모습을 보여줍니다.

위 정수인코딩 값중 최댓값이 7이라서 0 ~ 7까지 8가지 클래스로 표현하는 원핫인코딩결과

"""
None

## PYTorch의 `one_hot()`

In [142]:
import torch.nn.functional as F

In [145]:
# 예시 레이블 (정수 형태)
y = torch.tensor([0, 2, 1, 7])

# (4,) -> 원핫 -> (4, 8)

y_onehot = F.one_hot(y, num_classes=8)
y_onehot

tensor([[1, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 1]])

In [146]:
# 모델 입력으로 사용할 경우에는 float으로 변환
y_onehot = F.one_hot(y, num_classes=8).float()

y_onehot

tensor([[1., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 1., 0., 0., 0., 0., 0.],
        [0., 1., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 1.]])

##  원-핫 인코딩(One-Hot Encoding)의 한계
- 저장공간의 문제
- 단어의 유사도 표현 못함


In [ ]:
# 단어의 유사도?

# 웹 검색창에 '오사카 숙소' 라고 검색
#   '호텔', '게스트 하우스', '희준이네 집'

In [ ]:
# 단어의 유사도 문제 해결 -> 단어를 다차원 공간에 벡터화 기법 두가지
# 1. 카운트 기반의 벡터화 : ex) LSA(잠재 의미 분석), HAL
# 2. 예측 기반으로 벡터화 : NNLM, RNNLM, Word2Vec, FastText
#  카운트 기반과 예측 기반 두 가지 방법을 모두 사용 : ex) GloVe

# ■ Embedding

![](https://heung-bae-lee.github.io/image/why_is_sparse_matirx_one_hot_encoding.png)

![](https://heung-bae-lee.github.io/image/Dense_representation.png)

<br>

**Embedding Layer**

![](https://www.researchgate.net/profile/Manuel-Lopez-Martin/publication/349630764/figure/fig3/AS:999014610788354@1615195052671/Detail-of-the-embedding-layer-of-the-NN-implementing-the-Encoding-model-used-for-the.png)


In [ ]:
# 자연어 처리(Natural Language Processing)분야에서 임베딩(Embedding)은
# 사람이 쓰는 자연어를 기계가 이해할 수 있는 숫자형태인 vector로 바꾼 결과 혹은 그 일련의 과정 전체를 의미한다.

## PyTorch의 nn.Embedding()

**torch.nn.Embedding**
https://docs.pytorch.org/docs/stable/generated/torch.nn.Embedding.html

```python
class torch.nn.Embedding(
  num_embeddings,  # ex) 사전의 크기(입력 차원)
  embedding_dim,  # 임베딩 크기(출력 차원)
  padding_idx=None,
  max_norm=None,
  norm_type=2.0,
  scale_grad_by_freq=False,
  sparse=False,
  _weight=None,
  _freeze=False,
  device=None,
  dtype=None)
```



In [2]:
import torch
import torch.nn as nn

# 1️⃣ 단어 사전 크기(vocab_size)와 임베딩 차원 설정
vocab_size = 10    # 단어가 10개 있다고 가정 (정수 0~9)
embedding_dim = 4  # 각 단어를 4차원 벡터로 표현

# 2️⃣ Embedding 레이어 생성
embedding = nn.Embedding(num_embeddings=vocab_size, embedding_dim=embedding_dim)

# 3️⃣ 입력으로 단어 인덱스(정수)를 전달
# 예: 문장 [2, 5, 7] 을 임베딩에 통과
input_indices = torch.tensor([2, 5, 7])  # 정수형 텐서 (예: [2, 5, 7])

# 4️⃣ 임베딩 결과 출력
embedded = embedding(input_indices)

print("입력 인덱스:", input_indices)
print("임베딩 결과 텐서 크기:", embedded.shape)
  # 결과 텐서 크기 [입력길이, embedding_dim]  즉, [3, 4] 형태
print("임베딩 벡터:\n", embedded)

# 출력 예시 (랜덤 초기화됨)


입력 인덱스: tensor([2, 5, 7])
임베딩 결과 텐서 크기: torch.Size([3, 4])
임베딩 벡터:
 tensor([[-0.2271, -0.8260, -0.2969,  0.0178],
        [ 0.2391,  0.3225, -0.9725, -0.6240],
        [ 1.2212, -0.8302,  0.1404, -1.1139]], grad_fn=<EmbeddingBackward0>)


In [3]:
# 문장을 여러 단어로 구성된 배치(batch) 로 처리할 수도 있습니다.

# batch_size = 2, 문장 길이 = 3

input_batch = torch.tensor([[1, 2, 3],
                            [4, 5, 6]])

embedded_batch = embedding(input_batch)  # (2, 3) -> (2, 3, 4)

embedded_batch.shape  # 배치 2개, 각 문장(시퀀스) 3단어, 임베딩 차원 4

torch.Size([2, 3, 4])

In [ ]:
# nn.Embedding 역할
# 정수 인덱스를 실수 벡터로 변환
# 입력  LongTensor (ex: [1, 2, 3])
# 출력  FloatTensor (ex: [3, embedding_dim])
# 학습  다른 신경망 층과 함께 학습 가능 (임베딩 벡터는 requires_grad=True)

## keras 의 Embedding

**tf.keras.layers.Embedding**

https://www.tensorflow.org/api_docs/python/tf/keras/layers/Embedding

```python
tf.keras.layers.Embedding(
    input_dim,
    output_dim,
    embeddings_initializer='uniform',
    embeddings_regularizer=None,
    embeddings_constraint=None,
    mask_zero=False,
    weights=None,
    lora_rank=None,
    **kwargs
)
```

In [ ]:
# 이 레이어는 PyTorch의 nn.Embedding과 거의 동일한 역할을 함
# — 정수 인덱스를 실수 벡터(embedding) 로 변환해주는 레이어.

In [4]:
import tensorflow as tf

# 1️⃣ 단어 사전 크기(vocab_size)와 임베딩 차원 설정
vocab_size = 10     # 단어가 10개 (0~9)
embedding_dim = 4   # 각 단어를 4차원 벡터로 표현

# 2️⃣ Embedding 레이어 생성
embedding = tf.keras.layers.Embedding(input_dim=vocab_size, output_dim=embedding_dim)

# 3️⃣ 입력 (정수 인덱스)
# 문장: [2, 5, 7]
input_indices = tf.constant([2, 5, 7])

# 4️⃣ 임베딩 벡터 조회
embedded = embedding(input_indices)

print("입력 인덱스:", input_indices.numpy())
print("임베딩 결과 텐서 shape:", embedded.shape)  # (3, 4)
print("임베딩 벡터:\n", embedded.numpy())
# 출력 예시 (매번 달라짐)


입력 인덱스: [2 5 7]
임베딩 결과 텐서 shape: (3, 4)
임베딩 벡터:
 [[-0.01624779 -0.02526832  0.02469412 -0.03285334]
 [ 0.01653332  0.03077916  0.01185491 -0.02679151]
 [-0.04076462 -0.0494668   0.04941997 -0.02504973]]


In [6]:
# Embedding 레이어는 내부적으로 (vocab_size, embedding_dim) 크기의 룩업 테이블(행렬) 하나를 weight로 가지고 있습니다
embedding.get_weights()

[array([[ 0.0346132 , -0.00115231,  0.01597664, -0.04120301],
        [-0.0397033 , -0.00550023,  0.04593176,  0.03311844],
        [-0.01624779, -0.02526832,  0.02469412, -0.03285334],
        [-0.02893387,  0.03234618,  0.00312296, -0.02527386],
        [ 0.03076946, -0.03894892,  0.01231214,  0.02624087],
        [ 0.01653332,  0.03077916,  0.01185491, -0.02679151],
        [ 0.03649248, -0.02760246,  0.04338269, -0.03978806],
        [-0.04076462, -0.0494668 ,  0.04941997, -0.02504973],
        [-0.02141291, -0.00523871, -0.00937859, -0.0222013 ],
        [ 0.02800841, -0.03629438, -0.02634969, -0.00722054]],
       dtype=float32)]